### PR TITLE
Openemr swagger continued for api documentation

### PR DESCRIPTION
**Pull request by bradymiller**
This is a continuation from @adunsulag 's work at https://github.com/openemr/openemr/pull/4602

Plan:
1. Ensure I have a handle on all the scopes (first gonna get them right in API_README.md and will also incorporate them into the documentation in _rest_routes.inc.php).
2. **(DONE)** Ensure set things up to be able to easily test endpoints from swagger
3. Flush out all the documentation in _rest_routes.inc.php
4. Create a openemr-cmd in easy docker that runs `php bin/command-runner -c CreateAPIDocumentation` and that also registers a oauth2 client (and will return the client id and secret) (also make a command to do the same for a demo on the demo farm in case the dev wishes to test api's on a demo); place raw scopes_listing in `docker/library/api-scope-listing` so easy dev can use this when request the client registration (so do not need to keep rebuilding docker when add more scopes). Then place documentation in contrib for easy dev docker for using swagger etc.
5. **(DONE)** Get it working from the demo farm

demo for this PR is here:
https://www.open-emr.org/wiki/index.php/Development_Demo#Delta_-_Up_For_Grabs_Demo

### CODE REVIEW

File: swagger/openemr-api.json [hunk 1]
```diff
@@ -0,0 +1,384 @@
+{
+    "openapi": "3.0.0",
+    "info": {
+        "title": "OpenEMR API",
+        "version": "6.0.0"
```
Comment (stephenwaite) [inline swagger/openemr-api.json:5]: hi @bradymiller , is the version `6.1.0`?

Comment (bradymiller) [reply swagger/openemr-api.json:5]: yep, i thought i changes that, but will make sure (this PR is still a ways off; still trying to get the oauth2 process to work from within swagger to allow easy testing of the api and then will full build out the api documentation)

Comment (stephenwaite) [reply swagger/openemr-api.json:5]: really cool development, thank you @adunsulag & @bradymiller

Comment (bradymiller) [reply swagger/openemr-api.json:5]: oh actually, need to bring in most recent openemr-api.json/openemr-api.yaml (forgot to do that) to show most recent changes in the documentation.

**Comment by bradymiller** [comment]
ok, this is just too cool. Can now basically auth and run the api endpoints through the swagger ui!
First, register a client in postman:
image

Then in swagger ui (https://localhost:9300/swagger/), click Authorize (enter in the client_id and client_secret that get from above client registration and select all the scopes, then go through the auth and voila, you can then open up endpoints and run them through the swagger ui)

2DV

btw, still got a bunch more api documentation to do, but now have a nice starting point for testing and completing the documentation.

**Comment by bradymiller** [comment]
more cool stuff! Can now try it out on the demo farm. Here is the current up for grab demo that can run this from:
https://www.open-emr.org/wiki/index.php/Development_Demo#Delta_-_Up_For_Grabs_Demo

Basically do same steps as above in the easy dev docker.

First, register a client in postman:
image
(below is the text used in the body above):
```
   {
   "application_type": "private",
   "redirect_uris":
     ["https://four.openemr.io/a/openemr/swagger/oauth2-redirect.html"],
   "client_name": "A Private App",
   "token_endpoint_auth_method": "client_secret_post",
   "contacts": ["me@example.org", "them@example.org"],
   "scope": "openid offline_access api:fhir api:oemr api:port patient/AllergyIntolerance.read patient/CarePlan.read patient/CareTeam.read patient/Condition.read patient/Device.read patient/DiagnosticReport.read patient/DocumentReference.read patient/Encounter.read patient/Goal.read patient/Immunization.read patient/Location.read patient/Medication.read patient/MedicationRequest.read patient/Observation.read patient/Organization.read patient/Patient.read patient/Person.read patient/Practitioner.read patient/Procedure.read patient/Provenance.read user/AllergyIntolerance.read user/CarePlan.read user/CareTeam.read user/Condition.read user/Coverage.read user/Device.read user/DiagnosticReport.read user/DocumentReference.read user/Encounter.read user/Goal.read user/Immunization.read user/Location.read user/Medication.read user/MedicationRequest.read user/Observation.read user/Organization.read user/Organization.write user/Patient.read user/Patient.write user/Person.read user/Practitioner.read user/Practitioner.write user/PractitionerRole.read user/Procedure.read user/Provenance.read user/allergy.read user/allergy.write user/appointment.read user/appointment.write user/dental_issue.read user/dental_issue.write user/document.read user/document.write user/drug.read user/encounter.read user/encounter.write user/facility.read user/facility.write user/immunization.read user/insurance.read user/insurance.write user/insurance_company.read user/insurance_company.write user/insurance_type.read user/list.read user/medical_problem.read user/medical_problem.write user/medication.read user/medication.write user/message.write user/patient.read user/patient.write user/practitioner.read user/practitioner.write user/prescription.read user/procedure.read user/soap_note.read user/soap_note.write user/surgery.read user/surgery.write user/vital.read user/vital.write patient/encounter.read patient/patient.read"
  }
```

Then in swagger ui (https://four.openemr.io/a/openemr/swagger/index.html), click Authorize (enter in the client_id and client_secret that get from above client registration and select all the scopes, then go through the auth and voila, you can then open up endpoints and run them through the swagger ui).

The future development demos on the demo farm will all support this when this code is in the main codebase and plan to place a link to this in the demos titled something like "API (Swagger) User Interface:", see below for example of the link in the current up for grabs demo:
image

**Comment by bradymiller** [comment]
Testing well and going to bring this in.

Things that plan to do in very near future in another PR to both complete the documentation and make testing (in combination with flex docker updates) of the api more accessible:
1. Ensure I have a handle on all the scopes (first gonna get them right in API_README.md and will also incorporate them into the documentation in _rest_routes.inc.php).
2. Flush out all the documentation in _rest_routes.inc.php
3. Create a openemr-cmd in easy docker that runs `php bin/command-runner -c CreateAPIDocumentation` and that also registers a oauth2 client (and will return the client id and secret) (also make a command to do the same for a demo on the demo farm in case the dev wishes to test api's on a demo); place raw scopes_listing in `docker/library/api-scope-listing` so easy dev can use this when request the client registration (so do not need to keep rebuilding docker when add more scopes). Then place documentation in contrib for easy dev docker for using swagger etc.

**Comment by stephenwaite** [comment]
jarosiewicz-mr-bean

**Comment by bradymiller** [comment]
Number 3 done here:
https://github.com/openemr/openemr/pull/4611

Now just 1 and 2 to do :)

**Comment by bradymiller** [comment]
Just a cool thing related to item number 3:
Can now use swagger on the development demos. For example, if want to use swagger on the demo here:
https://www.open-emr.org/wiki/index.php/Development_Demo#Demo_.28Alpine_3.13_-_uses_php_8.0.29
(we'll set it up on the one without the demo data, just to show that it works for demos with a letter in the path)

via the most recent openemr-cmd, can do:
```
openemr-cmd rocd https://four.openemr.io/a/openemr
```
This will set up a oauth2 client to work with swagger (which is linked to on the demo link above), and the client id/secret that you got from above command can then be used there to Authorize in the swagger gui.

**Comment by bradymiller** [comment]
Just added devtool/openemr-cmd to support testing with swagger on multisites. Flow could look something like this.
Create some multisites that clone from default (below will create 3 multisites called run1, run2, run3):
```
openemr-cmd gmb 3
```
set up swagger to use the run1 multsite:
```
openemr-cmd swtm run1
```
set up the auth2 client for the run1 multisite:
```
openemr-cmd rocd run1
```

**Comment by adunsulag** [comment]
Great work on all of this @bradymiller!

**Pull request merged**